r/R/type.R

```diff
@@ -55,6 +59,12 @@ DataType$import_from_c <- ImportType
 INTEGER_TYPES <- as.character(outer(c("uint", "int"), c(8, 16, 32, 64), paste0))
 FLOAT_TYPES <- c("float16", "float32", "float64", "halffloat", "float", "double")
 
+code_carefully <- function(type, msg) {
```

Comment:
What does this do? And why is it only used in some places where `code` is called? (I'm assuming there are good answers to both, and the answers should go in code comments in/around this function.)

Reply:
The idea is to support things like this: 


``` r
library(arrow, warn.conflicts = FALSE)
#> See arrow_info() for available features

tryCatch(
  error=print, 
  struct(a = arrow:::DayTimeInterval__initialize())$code()
)
#> <error/rlang_error>
#> Error in `code()`:
#>   Problem getting code for field "a".
#> Caused by error in `code()`:
#>   Unsupported type: <day_time_interval>.
#>   ℹ The arrow package currently has no function to make these.
```

<sup>Created on 2021-12-03 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

instead of: 

``` r
library(arrow, warn.conflicts = FALSE)
#> See arrow_info() for available features

tryCatch(
  error=print, 
  struct(a = arrow:::DayTimeInterval__initialize())$code()
)
#> <error/rlang_error>
#> Error in `code()`: Unsupported type: <day_time_interval>.
#> ℹ The arrow package currently has no function to make these.
```

<sup>Created on 2021-12-03 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

but maybe that's overkill.

Reply:
Another way could be to always emit code, but inject call to `stop()` or `abort()` on unsupported types, i.e. we'd get: 

```r
struct(a = stop("Unsupported type: <day_time_interval>"))
```

This might be more useful to the user of `$code()`

Reply:
We can revert the last commit if needed, but I believe this is more useful, because at least. the user has the rest of the structure of the schema: 

``` r
library(arrow, warn.conflicts = FALSE)
#> See arrow_info() for available features

s <- struct(a = list_of(
  struct(b = arrow:::DayTimeInterval__initialize())
))

s$code()
#> struct(a = list_of(struct(b = stop("Unsupported type: <day_time_interval>."))))
```

<sup>Created on 2021-12-03 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>